src/factorizations/blocklanczos.jl

```diff
@@ -126,38 +124,36 @@ Here, [`initialize(::KrylovIterator)`](@ref) produces the first Krylov factoriza
 and [`expand!(iter::KrylovIterator, fact::KrylovFactorization)`](@ref) expands the
 factorization in place.
 """
-struct BlockLanczosIterator{F,T,S,O<:Orthogonalizer} <: KrylovIterator{F,T}
+struct BlockLanczosIterator{F,T,O<:Orthogonalizer,S<:Real} <: KrylovIterator{F,T}
     operator::F
-    x₀::Block{T,S}
+    x₀::Block{T}
     maxdim::Int
     orth::O
-    qr_tol::Real
-    function BlockLanczosIterator{F,T,S,O}(operator::F,
-                                           x₀::Block{T,S},
+    qr_tol::S
+    function BlockLanczosIterator{F,T,O,S}(operator::F,
+                                           x₀::Block{T},
                                            maxdim::Int,
                                            orth::O,
-                                           qr_tol::Real) where {F,T,S,O<:Orthogonalizer}
-        return new{F,T,S,O}(operator, x₀, maxdim, orth, qr_tol)
+                                           qr_tol::S) where {F,T,O<:Orthogonalizer,S<:Real}
+        return new{F,T,O,S}(operator, x₀, maxdim, orth, qr_tol)
     end
 end
 function BlockLanczosIterator(operator::F,
-                              x₀::Block{T,S},
+                              x₀::Block{T},
                               maxdim::Int,
-                              qr_tol::Real,
-                              orth::O=ModifiedGramSchmidt2()) where {F,T,S,
-                                                                     O<:Orthogonalizer}
+                              orth::O=KrylovDefaults.orth,
+                              qr_tol::Real=KrylovDefaults.tol) where {F,T,O<:Orthogonalizer}
     norm(x₀) < qr_tol && @error "initial vector should not have norm zero"
     orth != ModifiedGramSchmidt2() &&
         @error "BlockLanczosIterator only supports ModifiedGramSchmidt2 orthogonalizer"
-    return BlockLanczosIterator{F,T,S,O}(operator, x₀, maxdim, orth, qr_tol)
+    return BlockLanczosIterator{F,T,O,typeof(qr_tol)}(operator, x₀, maxdim, orth, qr_tol)
 end
 
-function initialize(iter::BlockLanczosIterator{F,T,S};
-                    verbosity::Int=KrylovDefaults.verbosity[]) where {F,T,S}
+function initialize(iter::BlockLanczosIterator;
+                    verbosity::Int=KrylovDefaults.verbosity[])
     X₀ = iter.x₀
     maxdim = iter.maxdim
     A = iter.operator
-    BTD = zeros(S, maxdim, maxdim)
 
     # Orthogonalization of the initial block
     X₁ = copy(X₀)
```

Comment:
Rather than a plain copy, here we should make sure that `X₁` is such that it supports in place `block_qr!`, i.e. it contains vectors with floating point scalar types. I will think some more about the best/nicest way to accomplish this.